### ISSUE TITLE
VACEngine is Abstract

### ISSUE THREAD

**Issue by Vencint** [dl1ycf/MacOSVirtualAudio#1]
When I try to compile the project in Xcode it says 

> Allocating an object of abstract class type 'VACEngine'

in `VACEngine.cpp` line 22 and `VACDevice.cpp` line 60.
It further says 

> 1. Unimplemented pure virtual method 'driverDesiresHiResSampleIntervals' in 'VACEngine'

so I added this line to the class declaration in `VACEngine.h`:
```c++
virtual bool driverDesiresHiResSampleIntervals(void);
```

I then wanted to install the kext via `kmutil load -p /Library/Extensions/VACDevice.kext` after which I have to approve in the Systems Preferences Security pane. I am then asked to reboot my computer but the kext is still not loaded afterwards. 🤨

Did I "break" the kernel extension by adding this new line? Could you please update your code, @dl1ycf, or tell me what I need to do in order to load the kext?

Thank you very much in advance! 🙏

**Comment by dl1ycf** [comment]
I know next to nothing about kernel extensions. I took an example
from a book and went on.

In principle, one needs to update the code for any new version of
MacOS since the interfaces slightly change. I did not look into
this because MacOS has announced stop of support for kernel extensions,
using IOkits or something like that in the future.

Frankly speaking, I shall use my kexts as long as they work (they work
for me with MacOS BigSur, but I have not tried to recompile them),
and then I will switch to a „Loopback" from RogueAmoeba that does the
same job and even more (this will add a patch panel).

So bottom line:

if it works, be happy.

**Comment by Vencint** [comment]
I see; thank you very much. :)

**Comment by dl1ycf** [comment]
If you use the „kextload“ and „kextutil“ commands, then you get more
meaningful error messages.

Be sure to DISABLE „system integrity protection“. Otherwise unsigned kexts
will never load. To disable sip, you have to boot into recovery mode,
enter the terminal and issue the command csrutil.

with „csrutil status“ you get the actual state, and with

„csrutil enable“ or „csrutil disable“ you enable/disable SIP. Then reboot.

I shall try to compile it again, whether it works, or whether one should
use an older version of XCode to do so.

Yours,

**Comment by dl1ycf** [comment]
Hi,

I just checked that the VAC driver builds on my MacOS using „BigSur“,
but note that I have installed XCocde 12.1 (you can still get it
from Apple’s developer site).

There are many warnings, but they all belong to one of two types:

- use of double-quotes in #include lines, this is in the Framework
  so does not matter

- use of functions marked „deprecated“. This I know, the VAC driver
  will cease to work on future versions of MacOS.


Then I verified that the kext loads with these steps:

- removed my old VACDevice.kext from /Library/Extensions and rebooted:
  VAC devices gone (SDR RX and SDR TX no longer visible in the AudioMidi
  program)

- IMPORTANT: in a terminal window, did a „sudo chown -R root:wheel“
  on the freshly compiled kext, all files *must* belong to root

- moved the new kext to /Library/Extensions and rebooted.

Result: kext not loaded. You have a consent to loading these kexts
in the „System preferences“ security panel. It is already there, you
simply have to hit „Agree“.

NOW COMES the weird thing, but this is a BigSur problem and has
nothing to do with the kext: MacOS BigSur „forgets“ the administrator
privileges of all users, so you cannot „unlock“ the „security“
system preferences panel which you have to in oder to apply a change.

From the internet, I got the following trick:

- log out, and in the login panel, choose „Other User“ and type in
„root“ for user and the root password. Then, you can „Agree“ in the
security panel and reboot et voila, the SDR TX and SDR RX „sound
cards“ have re-appeared in the audio-midi program.

So I can confirm that one can compile and install the loopback devices
with MacOS BigSur, but not with the lastet MacOS SDK and IoKit.
I haven’t tried to install the MacOS10.15 SDK in the most recent XCode app,
I find it simpler to „downgrade“ to XCode 12.1 which comes with the 
old SDK.


How to downgrade to XCode 12.1:

remove Xcode app from /Applications, get XCode 12.1 from Apple Developer
support site and install.

PLEASE, could you try this? If you can verify it works we should write
down the procedure and add this to the github REAMDE.

Yours,

Christoph DL1YCF.
